cdk/domino_cdk/config/base.py

```diff
@@ -0,0 +1,140 @@
+from dataclasses import dataclass, fields, is_dataclass
+from inspect import isclass
+from typing import Dict, List
+
+from ruamel.yaml.comments import CommentedMap
+
+from domino_cdk import __version__
+from domino_cdk.config.efs import EFS
+from domino_cdk.config.eks import EKS
+from domino_cdk.config.route53 import Route53
+from domino_cdk.config.s3 import S3
+from domino_cdk.config.vpc import VPC
+
+
+def from_loader(name: str, cfg, c: dict):
+    if c:
+        print(f"Warning: Unused/unsupported config entries in {name}: {c}")
+    return cfg
+
+
+@dataclass
+class MachineImage:
+    ami_id: str
+    user_data: str
+
+
+@dataclass
+class DominoCDKConfig:
+    schema: str
+    name: str
+    aws_region: str
+    aws_account_id: str
+    availability_zones: List[str]
+    tags: Dict[str, str]
+
+    vpc: VPC
+    efs: EFS
+    route53: Route53
+    eks: EKS
+    s3: S3
+
+    install: dict
+
+    @staticmethod
+    def from_0_0_0(c: dict):
+        return from_loader(
+            "config",
+            DominoCDKConfig(
+                schema=__version__,
+                name=c.pop("name"),
+                aws_region=c.pop("aws_region"),
+                aws_account_id=c.pop("aws_account_id"),
+                availability_zones=c.pop("availability_zones", []),
+                tags=c.pop("tags", {}),
+                install=c.pop("install", {}),
+                vpc=VPC.from_0_0_0(c.pop("vpc")),
+                efs=EFS.from_0_0_0(c.pop("efs")),
+                route53=Route53.from_0_0_0(c.pop("route53")),
+                eks=EKS.from_0_0_0(c.pop("eks")),
+                s3=S3.from_0_0_0(c.pop("s3")),
+            ),
+            c,
+        )
+
+    @staticmethod
+    def from_0_0_1(c: dict):
+        return from_loader(
+            "config",
+            DominoCDKConfig(
+                schema=__version__,
+                name=c.pop("name"),
+                aws_region=c.pop("aws_region"),
+                aws_account_id=c.pop("aws_account_id"),
+                availability_zones=c.pop("availability_zones", []),
+                tags=c.pop("tags", {}),
+                install=c.pop("install", {}),
+                vpc=VPC.from_0_0_1(c.pop("vpc")),
+                efs=EFS.from_0_0_0(c.pop("efs")),
+                route53=Route53.from_0_0_0(c.pop("route53")),
+                eks=EKS.from_0_0_1(c.pop("eks")),
+                s3=S3.from_0_0_0(c.pop("s3")),
+            ),
+            c,
+        )
+
+    def __post_init__(self):
+        errors = []
+
+        def val(path: str, obj):
```

Comment:
this is really cool :+1:

I wonder if generating a JSON schema or something similar from the dataclass itself could be a next step so we could defer the type checking validation earlier. On the other hand, this kind of construction would let us hook in custom validation through dataclass field `metadata` or something.